### ISSUE TITLE
layer_finder.job is nil in JobConverter which gives undefined method `legacy_metric_name`

### ISSUE THREAD

**Issue by jpslav** [scoutapp/scout_apm_ruby#145]
We just upgraded from 3.0.0.pre13 to 3.0.0.pre16 in four different repositories.  For some reason, one of the repos is giving us a bunch of `NoMethodError: undefined method `legacy_metric_name' for nil:NilClass` errors in our specs (https://travis-ci.org/openstax/tutor-server/jobs/325192670).  

With debugger found this is happening at https://github.com/scoutapp/scout_apm_ruby/blob/master/lib/scout_apm/layer_converters/job_converter.rb#L21 :

```ruby
@meta_options = {:scope => layer_finder.job.legacy_metric_name}
```

`layer_finder.job` is `nil`.  The `request.job?` check several lines up did return `true`.

Any ideas?

**Comment by cschneid** [comment]
It's odd we'd be running in test mode at all. Check that `monitor: false` is in the test part of scout_apm.yml.

Does the list of failing tests stay stable between runs?

I wonder if it's related to the thread-local "current-transaction" that is running. If it gets in a confused state, I could see it answering 'yes' to being a job, but not actually having the data inside that is required. You can clear any request by doing: `Thread.current[:scout_request] = nil`.

**Comment by jpslav** [comment]
@cschneid There isn't even a `scout_apm.yml` file in our repository -- we only write that config file to our deployed production servers.  (here's the repo https://github.com/openstax/tutor-server/tree/scout-apm-update)

I'm not sure if the failure list is consistent, but I am able to replicate the error over and over in any given test.  

I reverted to `3.0.0.pre13` and the spec passed.  I upgraded one step at a time and saw the failure when moving to `3.0.0.pre14`.

When I am in the failing spec, I get:

```ruby
(byebug) ScoutApm::Agent.instance.context.config.value("monitor")
false
```

Inside the failing spec, `Thread.current[:scout_request]` is definitely not nil, and inside the `TrackedRequest` object used in the `JobConverter`, I get 

```ruby
(byebug) @agent_context.config.value("monitor")
false
```

**Comment by cschneid** [comment]
@jpslav your debugging is amazing. If that repo reproduces the error, I'll fix it, and let you know. I'm sorry for the hassle.

**Comment by jpslav** [comment]
If it is non trivial to set up that repo, you can use https://github.com/openstax/lms-dev-env and comment out https://github.com/openstax/lms-dev-env/blob/master/main.yml#L3 to get our app installed in a Vagrant box.

**Comment by cschneid** [comment]
Got it going, and just reproduced what you were seeing - taking a look.

**Comment by cschneid** [comment]
@jpslav  it took some searching but I figured out what's happening.  A bit of background first:

Some instrumentation is marked as `ignore_children!`, so that inner instrumentation isn't run. For example, when instrumenting elastic search, the http instrumentation may get hit, but should just be rolled into the elastic search instrument, not called out as a separate item.

ActiveRecord uses something similar. We instrument AR at a few levels, and often, a single call will go through several layers of our instruments, and we don't want to start many distinct nested instrumentation points.

What happened in this case was that the tests ran something like `FactoryBot.create(...)`, which started one of these `ignoring_children` spans. As part of the ActiveRecord save, it ran an inline Delayed Job, which marked the whole transaction as part of a job. When the transaction finished, we went to record it, and the "converter" class sees that it was marked as a job, but without any of the job details (ignored), causing the exception.

There are a few things I'm going to fix:

* Instrumentation shouldn't have installed in test mode anyway (a bit more investigation needed here).
* Simplify the logic determining if a transaction was a job, or web transaction. 

Keep an eye out for the next release. Sorry for the inconvenience, this is a tricky one.

**Comment by jpslav** [comment]
Cool thanks for the investigating!

**Comment by jpslav** [comment]
Hi @cschneid - I see 2.4.3 has a commit talking about this issue.  We were on 3.0.0.pre16.  Is that line of work still around?  If so, can the fix be released on 3.x as well?

**Comment by cschneid** [comment]
@jpslav it was merged into the Scoutprof (3.0.x) branch, but not yet released. I just pushed it as `3.0.0.pre17`.

**Comment by jpslav** [comment]
Thanks @cschneid our specs passed with `3.0.0.pre17` release!

**Comment by dlanderson** [comment]
Resolved in `2.4.3` and `3.0.0.pre17`
